ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionClientEngineApiCapabilitiesProvider.java

```diff
@@ -0,0 +1,167 @@
+      Collections.synchronizedSet(new HashSet<>());
+  private final AtomicBoolean isExecutionClientAvailable = new AtomicBoolean(false);
+  private final AtomicInteger failedFetchCounter = new AtomicInteger(0);
+  private Cancellable fetchTask;
+
+  public ExecutionClientEngineApiCapabilitiesProvider(
+      final AsyncRunner asyncRunner,
+      final ExecutionEngineClient executionEngineClient,
+      final EngineApiCapabilitiesProvider localEngineApiCapabilitiesProvider,
+      final EventChannels eventChannels) {
+    this.asyncRunner = asyncRunner;
+    this.executionEngineClient = executionEngineClient;
+    this.localEngineApiCapabilitiesProvider = localEngineApiCapabilitiesProvider;
+
+    eventChannels.subscribe(ExecutionClientEventsChannel.class, this);
+  }
+
+  private Cancellable runFetchTask() {
+    LOG.trace("Exchange Capabilities Task - Starting...");
+
+    return asyncRunner.runWithFixedDelay(
+        this::fetchRemoteCapabilities,
+        Duration.ZERO,
+        Duration.ofSeconds(EXCHANGE_CAPABILITIES_FETCH_INTERVAL_IN_SECONDS),
+        this::handleFetchCapabilitiesUnexpectedError);
+  }
+
+  private void fetchRemoteCapabilities() {
+    if (!isExecutionClientAvailable.get()) {
+      return;
+    }
+
+    final SafeFuture<Response<List<String>>> exchangeCapabilitiesCall =
+        executionEngineClient.exchangeCapabilities(
+            new ArrayList<>(localEngineApiCapabilitiesProvider.supportedMethodsVersionedNames()));
+
+    exchangeCapabilitiesCall
+        .thenAccept(
+            response -> {
+              if (response.isSuccess()) {
+                handleSuccessfulResponse(response);
+              } else {
+                handleFailedResponse(response.getErrorMessage());
+              }
+            })
+        .ifExceptionGetsHereRaiseABug();
+  }
+
+  private void handleSuccessfulResponse(final Response<List<String>> response) {
+    LOG.trace("Handling successful response (response = {})", response.getPayload());
+
+    final List<String> remoteCapabilities =
+        response.getPayload() != null ? response.getPayload() : List.of();
+    remoteSupportedMethodNames.addAll(remoteCapabilities);
+    remoteSupportedMethodNames.removeIf(m -> !remoteCapabilities.contains(m));
+
+    failedFetchCounter.set(0);
+  }
+
+  private void handleFailedResponse(final String errorResponse) {
+    final int failedAttempts = failedFetchCounter.incrementAndGet();
+
+    final StringBuilder sb = new StringBuilder();
+    sb.append(
+        String.format(
+            "Error fetching remote capabilities from Execution Client (failed attempts = %d)",
+            failedAttempts));
+    if (StringUtils.isNotBlank(errorResponse)) {
+      sb.append(". ").append(errorResponse);
+    }
+
+    if (failedAttempts >= EXCHANGE_CAPABILITIES_ATTEMPTS_BEFORE_LOG_WARN) {
+      LOG.warn(sb.toString());
+    } else {
+      LOG.trace(sb.toString());
+    }
+
+    if (remoteSupportedMethodNames.isEmpty()) {
```

Comment:
assuming we can't have both these scenarios and get a warn and error together?

Reply:
it could  get noisy very quickly! we may make several calls in a few seconds and start completely spamming...

Reply:
We only do it regularly every 15 minutes. So you get a warning log on the 3rd failed attempt (45 minutes). After that, you get a warning every 15 minutes if your EL never recovered.

The error log only happens if the EL has never responded successfully. I can check using the same check `remoteSupportedMethodNames.isEmpty()` and only log trace/warn when it is not empty.